Gradle `quarkusRun` task starts in `TEST` mode

### Describe the bug

1eaecd624368f156ecaeb8c52debbbd0961e5229 (via #40273) introduced a [change](https://github.com/quarkusio/quarkus/pull/40273/changes#diff-a513c2e2e5e36f6b844d77846742f83db6d649f763ab9cdafa15804449f20c90) to the `quarkusRun` Gradle task to let it start in `TEST` mode. This triggers Arc's test-only build steps, but resolution only happens for a normal app. `io.quarkus.test.ActivateSessionContext` is then absent (it's in `quarkus-test-common`) and startup breaks with this message.

```
Execution failed for task ':polaris-server:quarkusRun' (registered by plugin 'io.quarkus').
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.BeanArchiveProcessor#build threw an exception: java.lang.IllegalStateException: Failed to index: io.quarkus.test.ActivateSessionContext
        at io.quarkus.deployment.index.IndexingUtil.indexClass(IndexingUtil.java:165)
        at io.quarkus.arc.deployment.BeanArchiveProcessor.build(BeanArchiveProcessor.java:97)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:898)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:242)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
  Caused by: java.lang.IllegalStateException: Failed to index: io.quarkus.test.ActivateSessionContext, class not present in class loader: QuarkusClassLoader:Deployment Class Loader: TEST for polaris-server-1.4.0-SNAPSHOT@45559b9f
        at io.quarkus.deployment.index.IndexingUtil.indexClass(IndexingUtil.java:156)
        ... 11 more
```

This change appears to be introduced in Quarkus 3.34.0.

For int-test-like behavior there's already `quarkusDev`.

(A "normal" `quarkusBuild` + `java -jar runtime/server/build/quarkus-app/quarkus-run.jar` is still fine - it's only the `quarkusRun` task. `quarkusDev` also appears to work fine.)

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusRun.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusRun.java
@@ -115,7 +115,7 @@ public abstract class QuarkusRun extends QuarkusBuildTask {
                 .setAppArtifact(appModel.getAppArtifact())
                 .setLocalProjectDiscovery(false)
                 .setIsolateDeployment(true)
-                .setMode(QuarkusBootstrap.Mode.TEST)
+                .setMode(QuarkusBootstrap.Mode.RUN)
                 .build().bootstrap()) {
 
             AugmentAction action = curatedApplication.createAugmentor();

--- a/integration-tests/gradle/pom.xml
+++ b/integration-tests/gradle/pom.xml
@@ -86,6 +86,10 @@
         <!-- START update-dependencies.sh -->
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-agroal</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-arc</artifactId>
         </dependency>
         <dependency>
@@ -115,6 +119,10 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-hibernate-orm-panache</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-rest-data-panache</artifactId>
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
@@ -158,6 +166,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-liquibase</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-reactive-pg-client</artifactId>
         </dependency>
         <dependency>
@@ -195,6 +207,19 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-vertx-http</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-agroal-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
@@ -289,6 +314,19 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-rest-data-panache-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-hibernate-reactive-panache-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
@@ -368,6 +406,19 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-kubernetes-client-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-liquibase-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
             <scope>test</scope>

--- a/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/build.gradle
+++ b/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/build.gradle
@@ -1,0 +1,28 @@
+plugins {
+    id 'java'
+    id 'io.quarkus'
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex 'io.quarkus.*'
+            includeGroup 'org.hibernate.orm'
+        }
+    }
+    mavenCentral()
+    gradlePluginPortal()
+}
+
+dependencies {
+    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
+    implementation 'io.quarkus:quarkus-arc'
+    implementation 'io.quarkus:quarkus-rest'
+    implementation 'io.quarkus:quarkus-hibernate-orm-rest-data-panache'
+    implementation 'io.quarkus:quarkus-jdbc-h2'
+    implementation 'io.quarkus:quarkus-agroal'
+    implementation 'io.quarkus:quarkus-liquibase'
+}
+
+group 'org.acme'
+version '1.0.0-SNAPSHOT'

--- a/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/gradle.properties
@@ -1,0 +1,2 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/settings.gradle
+++ b/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/settings.gradle
@@ -1,0 +1,16 @@
+pluginManagement {
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex 'io.quarkus.*'
+                includeGroup 'org.hibernate.orm'
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+      id 'io.quarkus' version "${quarkusPluginVersion}"
+    }
+}
+rootProject.name = 'quarkus-run-mode-panache-liquibase'

--- a/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/src/main/java/org/acme/GreetingResource.java
+++ b/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/src/main/java/org/acme/GreetingResource.java
@@ -1,0 +1,16 @@
+package org.acme;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+@Path("/hello")
+public class GreetingResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "Hello from Quarkus REST...";
+    }
+}

--- a/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/src/main/resources/application.properties
+++ b/integration-tests/gradle/src/main/resources/quarkus-run-mode-panache-liquibase/src/main/resources/application.properties
@@ -1,0 +1,4 @@
+quarkus.datasource.db-kind=h2
+quarkus.datasource.jdbc.url=jdbc:h2:mem:test
+quarkus.hibernate-orm.database.generation=none
+quarkus.liquibase.migrate-at-start=false

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/run/QuarkusRunWithRestDataPanacheAndLiquibaseTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/run/QuarkusRunWithRestDataPanacheAndLiquibaseTest.java
@@ -1,0 +1,29 @@
+package io.quarkus.gradle.run;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.quarkus.gradle.devmode.QuarkusDevGradleTestBase;
+
+// Reproduces https://github.com/quarkusio/quarkus/issues/54001.
+// Despite extending QuarkusDevGradleTestBase, this test does not use dev mode,
+// but rather runs the application in production mode using the `quarkusRun` task.
+// The combination of `hibernate-orm-rest-data-panache` and `liquibase` triggered
+// the regression introduced when `quarkusRun` was bootstrapping the app with
+// `QuarkusBootstrap.Mode.TEST` (see #48950 and `46ce666581d` for the Maven-side fix).
+public class QuarkusRunWithRestDataPanacheAndLiquibaseTest extends QuarkusDevGradleTestBase {
+
+    @Override
+    protected String projectDirectoryName() {
+        return "quarkus-run-mode-panache-liquibase";
+    }
+
+    @Override
+    protected String[] buildArguments() {
+        return new String[] { "clean", "quarkusRun" };
+    }
+
+    @Override
+    protected void testDevMode() throws Exception {
+        assertThat(getHttpResponse("/hello")).contains("Hello from Quarkus REST...");
+    }
+}

